### PR TITLE
[OpenAI] Add logprob support

### PR DESCRIPTION
**Pull request by dsikka**
# Summary
- Now that we can return logits, add `logprob` support 
- This adds the `LogProbs` pydantic model, which can now be returned for the `completions` endpoint
- This model includes the list of produced tokens as well as the list of logprob associated with each token 
- Also please pay attention to the logprob calculation 

# Testing / Question?

- This was finicky to test because we dont output the generated `token_ids`, just the final sequence. So this ends up going through the steps of tokenizing the output to get the `token_ids` (as needed by the `LogProbs`) which seems redundant to do as we have them from the pipeline, but it is not part of the output --> **we should make it as part of the output to clean this up if there are not strong opinions against it?**
- This would also make it easier to handle the streaming and non-streaming case, which have a nuanced differences when tokenizing individual tokens vs full sequences (which was very annoying)

Local Testing
```yaml
num_cores: 2
num_workers: 2
integration: openai
endpoints:
  - task: text_generation
    model: "hf:mgoin/TinyStories-1M-ds"

```

Client Code:

```python
import openai
from openai import OpenAI


client = OpenAI(base_url="http://localhost:5543/v1", api_key="EMPTY")

models = client.models.list()

model = "hf:mgoin/TinyStories-1M-ds"
print(f"Accessing model API '{model}'")


# Completion API
stream = True
completion = client.completions.create(
    prompt="The dog",
    max_tokens=10,
    stream=stream,
    model=model,
    logprobs=True
)

for c in completion:
    print(c)
```

Output: 

```bash
Accessing model API 'hf:mgoin/TinyStories-1M-ds'
Completion(id='cmpl-67b04e6e35a84efd9950438d47c1794c', choices=[CompletionChoice(finish_reason=None, index=None, logprobs=Logprobs(text_offset=[], token_logprobs=[-1.5180178880691528], tokens=[' very'], top_logprobs=None), text=' very')], created=1705011569, model='/home/dsikka/.cache/huggingface/hub/models--mgoin--TinyStories-1M-ds/snapshots/eba4f1f16d00041d78c8f8a3bba5d3fa5afe5513/model.onnx', object='text_completion', system_fingerprint=None, usage=None)
Completion(id='cmpl-67b04e6e35a84efd9950438d47c1794c', choices=[CompletionChoice(finish_reason=None, index=None, logprobs=Logprobs(text_offset=[], token_logprobs=[-1.396243691444397], tokens=[' scared'], top_logprobs=None), text=' scared')], created=1705011569, model='/home/dsikka/.cache/huggingface/hub/models--mgoin--TinyStories-1M-ds/snapshots/eba4f1f16d00041d78c8f8a3bba5d3fa5afe5513/model.onnx', object='text_completion', system_fingerprint=None, usage=None)
Completion(id='cmpl-67b04e6e35a84efd9950438d47c1794c', choices=[CompletionChoice(finish_reason=None, index=None, logprobs=Logprobs(text_offset=[], token_logprobs=[-0.6142183542251587], tokens=['.'], top_logprobs=None), text='.')], created=1705011569, model='/home/dsikka/.cache/huggingface/hub/models--mgoin--TinyStories-1M-ds/snapshots/eba4f1f16d00041d78c8f8a3bba5d3fa5afe5513/model.onnx', object='text_completion', system_fingerprint=None, usage=None)
Completion(id='cmpl-67b04e6e35a84efd9950438d47c1794c', choices=[CompletionChoice(finish_reason=None, index=None, logprobs=Logprobs(text_offset=[], token_logprobs=[-0.7276340126991272], tokens=[' He'], top_logprobs=None), text=' He')], created=1705011569, model='/home/dsikka/.cache/huggingface/hub/models--mgoin--TinyStories-1M-ds/snapshots/eba4f1f16d00041d78c8f8a3bba5d3fa5afe5513/model.onnx', object='text_completion', system_fingerprint=None, usage=None)
Completion(id='cmpl-67b04e6e35a84efd9950438d47c1794c', choices=[CompletionChoice(finish_reason=None, index=None, logprobs=Logprobs(text_offset=[], token_logprobs=[-2.1691243648529053], tokens=[' wanted'], top_logprobs=None), text=' wanted')], created=1705011569, model='/home/dsikka/.cache/huggingface/hub/models--mgoin--TinyStories-1M-ds/snapshots/eba4f1f16d00041d78c8f8a3bba5d3fa5afe5513/model.onnx', object='text_completion', system_fingerprint=None, usage=None)
Completion(id='cmpl-67b04e6e35a84efd9950438d47c1794c', choices=[CompletionChoice(finish_reason=None, index=None, logprobs=Logprobs(text_offset=[], token_logprobs=[-0.06524639576673508], tokens=[' to'], top_logprobs=None), text=' to')], created=1705011569, model='/home/dsikka/.cache/huggingface/hub/models--mgoin--TinyStories-1M-ds/snapshots/eba4f1f16d00041d78c8f8a3bba5d3fa5afe5513/model.onnx', object='text_completion', system_fingerprint=None, usage=None)
Completion(id='cmpl-67b04e6e35a84efd9950438d47c1794c', choices=[CompletionChoice(finish_reason=None, index=None, logprobs=Logprobs(text_offset=[], token_logprobs=[-2.3006293773651123], tokens=[' go'], top_logprobs=None), text=' go')], created=1705011569, model='/home/dsikka/.cache/huggingface/hub/models--mgoin--TinyStories-1M-ds/snapshots/eba4f1f16d00041d78c8f8a3bba5d3fa5afe5513/model.onnx', object='text_completion', system_fingerprint=None, usage=None)
Completion(id='cmpl-67b04e6e35a84efd9950438d47c1794c', choices=[CompletionChoice(finish_reason=None, index=None, logprobs=Logprobs(text_offset=[], token_logprobs=[-1.681646466255188], tokens=[' back'], top_logprobs=None), text=' back')], created=1705011569, model='/home/dsikka/.cache/huggingface/hub/models--mgoin--TinyStories-1M-ds/snapshots/eba4f1f16d00041d78c8f8a3bba5d3fa5afe5513/model.onnx', object='text_completion', system_fingerprint=None, usage=None)
Completion(id='cmpl-67b04e6e35a84efd9950438d47c1794c', choices=[CompletionChoice(finish_reason=None, index=None, logprobs=Logprobs(text_offset=[], token_logprobs=[-0.7840849757194519], tokens=[' to'], top_logprobs=None), text=' to')], created=1705011569, model='/home/dsikka/.cache/huggingface/hub/models--mgoin--TinyStories-1M-ds/snapshots/eba4f1f16d00041d78c8f8a3bba5d3fa5afe5513/model.onnx', object='text_completion', system_fingerprint=None, usage=None)
```

### CODE REVIEW

File: src/deepsparse/server/openai_server.py [hunk 1]
```diff
@@ -431,6 +450,22 @@ def tokenize(text: str) -> List[int]:
                 )
 
 
+def create_logprobs(
+    token_ids: List[int], log_probs: numpy.ndarray, pipeline: Pipeline
+) -> LogProbs:
+
+    logprobs = LogProbs()
```
Comment (bfineran) [inline src/deepsparse/server/openai_server.py:457]: let's change the names so we don't have `logprobs` and `log_probs`

**Comment by dsikka** [comment]
This is ready for merge apart from support for the field`top_logprob`.
Need clarification on what we're actually returning for this field

**Review by dbogunowicz — DISMISSED** [review]

File: src/deepsparse/server/openai_server.py [hunk 2]
```diff
@@ -194,7 +198,7 @@ async def create_chat_completion(raw_request: Request):
                 )
                 choices.append(choice_data)
 
-            num_prompt_tokens = len(final_res.prompt_token_ids["input_ids"])
+            num_prompt_tokens = len(final_res.prompt_token_ids["input_ids"][0])
```
Comment (mgoin) [inline src/deepsparse/server/openai_server.py:201]: `prompt_token_ids` is now a List[List[int]]? maybe we should note that batching isn't supported and this is why `[0]` is valid

File: src/deepsparse/server/protocol.py [hunk 3]
```diff
@@ -142,17 +139,14 @@ class CompletionRequest(BaseModel):
...
     stop: Optional[Union[str, List[str]]] = Field(default_factory=list)
     presence_penalty: Optional[float] = 0.0
     frequency_penalty: Optional[float] = 0.0
-    best_of: Optional[int] = None
     logit_bias: Optional[Dict[str, float]] = None
     user: Optional[str] = None
     # Additional parameters
     top_k: Optional[int] = -1
     ignore_eos: Optional[bool] = False
-    use_beam_search: Optional[bool] = False
```
Comment (mgoin) [inline src/deepsparse/server/protocol.py:155]: nit: i'm not sure if the error will be informative for the user if these arguments are provided and the parsing fails, rather than having explicit errors that they aren't supported once we parse the args in the request function

**Review by mgoin — COMMENTED** [review]
could you add a test to compare the openai completions logprob with the pipeline score?

File: src/deepsparse/server/protocol.py (hunk 3 above)
Comment (dsikka) [reply src/deepsparse/server/protocol.py:155]: I usually look at api specs to see what is supported (at least when using other 3rd party tools in the past). I guess that would be openai docs which would be misleading, unless we had specific ones for deepsparse?

File: src/deepsparse/server/openai_server.py (hunk 2 above)
Comment (dsikka) [reply src/deepsparse/server/openai_server.py:201]: it's the format being returned by the pipeline but yah, I can make note.

**Comment by dsikka** [comment]
@mgoin When you're happy with this and it passes any tests you have, let me know and we can merge this in.
Also, if the tests you're running locally are worth adding to the openai_server tests, I can do that as well as part of this PR

File: setup.py [hunk 4]
```diff
@@ -107,6 +107,7 @@ def _parse_requirements_file(file_path):
     "flask>=1.0.0",
     "flask-cors>=3.0.0",
     "Pillow>=8.3.2",
+    "openai",
```
[the remark this answers is not in the document]
Comment (dsikka) [reply setup.py:110]: ah it's only needed for the client. so should still be in dev? @mgoin

[the remark this answers is not in the document]
Comment (mgoin) [reply setup.py:110]: let's just keep it in dev, it's not that bad of a dep

**Review by mgoin — APPROVED** [review]
